packages/toolkit/src/query/react/buildHooks.ts

```diff
@@ -717,47 +722,37 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
         Definitions
       >
       const dispatch = useDispatch<ThunkDispatch<any, any, AnyAction>>()
-      const [requestId, setRequestId] = useState<string>()
-
-      const promiseRef = useRef<MutationActionCreatorResult<any>>()
+      const [promise, setPromise] = useState<MutationActionCreatorResult<any>>()
 
-      useEffect(() => {
-        return () => {
-          promiseRef.current?.unsubscribe()
-          promiseRef.current = undefined
-        }
-      }, [])
+      useEffect(() => () => promise?.unsubscribe(), [promise])
 
       const triggerMutation = useCallback(
         function (arg) {
-          let promise: MutationActionCreatorResult<any>
-          batch(() => {
-            promiseRef?.current?.unsubscribe()
-            promise = dispatch(initiate(arg))
-            promiseRef.current = promise
-            setRequestId(promise.requestId)
-          })
-          return promise!
+          const promise = dispatch(initiate(arg))
+          setPromise(promise)
+          return promise
         },
         [dispatch, initiate]
       )
 
       const mutationSelector = useMemo(
         () =>
-          createSelector([select(requestId || skipToken)], (subState) =>
-            selectFromResult(subState)
+          createSelector(
+            [select(promise?.requestId || skipToken)],
+            selectFromResult
           ),
-        [select, requestId, selectFromResult]
+        [select, promise, selectFromResult]
       )
 
       const currentState = useSelector(mutationSelector, shallowEqual)
-      const originalArgs = promiseRef.current?.arg.originalArgs
+      const originalArgs = promise?.arg.originalArgs
       const finalState = useMemo(
         () => ({
           ...currentState,
           originalArgs,
+          reset: promise?.unsubscribe,
```

Comment:
```suggestion
          reset: promise?.unsubscribe ?? (() => undefined),
```

As it is, this results in lying to the types: Before any mutation is triggered, `reset` will be undefined rather than `() => void`, and hence result in a `TypeError` if attempted to be called. It's probably better to let it do a no-op call in that situation

Reply:
Very good catch, I gotta fix that one!

Reply:
❓ 

Just a bit of a lingering thought:

Given that the `reset` method is essentially an alias to the latest `promise.unsubscribe` property, it may come across as having counter-intuitive behaviour from the user's perspective.

i.e. a user will not intuitively expect that they *need* to be concerned about closures over `reset`, and that they *have to* access the latest one to get the desired 'resetting' behaviour.

For example, the below will *never* work from the user's perspective, as `result.reset` will be for the old, already-unsubscribed promise at that point, and won't unsubscribe from the fresh trigger.  The behaviour will appear as though `reset()` was never called at all.
```
  const [trigger, result] = useEditFooMutation();

  async function mutateAndResetAfter1s() {
    trigger("mutateAndResetAfter1s");
    await new Promise((r) => setTimeout(r, 1000));
    result.reset();
  }
```

An option to overcome this could be to use a stable 'reset' function that itself refers to the latest `promise.unsubscribe` via a ref, but I'd like to hear your thoughts on this

Reply:
I get where you're coming from. In that case it might be sensible to move it out of the `result` though to make clear you are not resetting "that result". (and here I had hoped I had gotten rid of all refs)

```
const [trigger, result, {reset}] = useEditFooMutation();
```

Reply:
Yeah, I can see it going either way really. What I'm anticipating is that if it's left as-is, even with `reset` being attached to the `result` object, people will overlook that as an 'implementation quirk', as their goal is to 'reset the hook', rather than 'reset a specific result'.

I'm honestly not sure which of the two options I prefer personally. It would be nice to avoid additional elements in the returned tuple, as well as avoiding the ref usage. And the hypothetical above might not ever come up in a real situation anyway.

For the particular situation above, we could also consider having another `reset` attached to the promise returned by `trigger()` itself, and go all-in on 'reset is for a particular result, not the hook in general'

Reply:
> For the particular situation above, we could also consider having another reset attached to the promise returned by trigger() itself, and go all-in on 'reset is for a particular result, not the hook in general'

Actually would `triggerReturn.unsubscribe()` already handle that?

Reply:
Yes it would.
We could also rename this `reset` function to `unsubscribe` and just stay with one naming

Reply:
Yeah I like the sound of that, it should prevent any confusion

Reply:
Hmm. On second thought, that could also add more confusion in combination with #1477.

My idea is there that you can have multiple components using the same mutation result by manually specifying a cache key:

```
const [trigger, result] = useEditFooMutation({ fixedCacheKey: "fooBar" });
```

Now, `result.unsubscribe()` would be a weird way of saying "reset that mutation result for all of those components".

Maybe instead I should rename `unsubscribe` to `reset` for mutations in general?

Reply:
Ah, that's a good point.

> Maybe instead I should rename unsubscribe to reset for mutations in general?

Yeah, at the moment that sounds like the most consistent outcome.

Reply:
Then let's think about wording here for a minute. 

* ~~unsubscribe~~
* reset
* removeFromCache
* remove
* ...more ideas?

Reply:
The only others I can currently come up with:  
- restore
- restoreCache

But I prefer just `reset` out of all of those options.

Reply:
`nukeItAll`?  `terminateWithExtremePrejudice`? :)